test/extended/router/h2spec.go

```diff
@@ -0,0 +1,289 @@
+package router
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+	"time"
+
+	g "github.com/onsi/ginkgo"
+	o "github.com/onsi/gomega"
+
+	routeclientset "github.com/openshift/client-go/route/clientset/versioned"
+	"github.com/openshift/origin/test/extended/router/h2spec"
+	exutil "github.com/openshift/origin/test/extended/util"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	e2e "k8s.io/kubernetes/test/e2e/framework"
+)
+
+type h2specFailingTest struct {
+	TestCase   *h2spec.JUnitTestCase
+	TestNumber int
+}
+
+type h2specRouteTypeTest struct {
+	routeType     string
+	hostname      string
+	knownFailures map[string]bool
+}
+
+func (f h2specFailingTest) ID() string {
+	return fmt.Sprintf("%s.%d", f.TestCase.Package, f.TestNumber)
+}
+
+func (t h2specRouteTypeTest) h2specCommand(results string) string {
+	return fmt.Sprintf("h2spec --tls --insecure --strict --host=%s --junit-report=%s", t.hostname, results)
+}
+
+var _ = g.Describe("[sig-network-edge][Conformance][Area:Networking][Feature:Router]", func() {
+	defer g.GinkgoRecover()
+
+	var (
+		configPath = exutil.FixturePath("testdata", "router", "router-h2spec.yaml")
+		oc         = exutil.NewCLI("router-h2spec", exutil.KubeConfigPath())
+	)
+
+	// this hook must be registered before the framework namespace teardown
+	// hook
+	g.AfterEach(func() {
+		if g.CurrentGinkgoTestDescription().Failed {
+			client := routeclientset.NewForConfigOrDie(oc.AdminConfig()).RouteV1().Routes(oc.KubeFramework().Namespace.Name)
+			if routes, _ := client.List(metav1.ListOptions{}); routes != nil {
+				outputIngress(routes.Items...)
+			}
+			exutil.DumpPodLogsStartingWith("h2spec", oc)
+		}
+	})
+
+	g.By(fmt.Sprintf("creating routes from a config file %q", configPath))
+
+	g.Describe("The HAProxy router", func() {
+		g.It("should pass the h2spec conformance tests", func() {
+			var err error
+
+			_, err = exutil.WaitForRouterServiceIP(oc)
+			o.Expect(err).NotTo(o.HaveOccurred())
+
+			err = oc.Run("create").Args("-f", configPath).Execute()
+			o.Expect(err).NotTo(o.HaveOccurred())
```

Comment:
Are you tearing down the test namespace anywhere?

Reply:
I believe so:

```console
STEP: [reencrypt] Fetching "/tmp/runConformanceTests912928811" from the container
Mar 18 12:30:07.021: INFO: Running 'oc --namespace=e2e-test-router-h2spec-8dl5l --kubeconfig=/tmp/configfile241935612 rsync h2spec:/tmp/runConformanceTests912928811 /tmp --strategy=tar'
runConformanceTests912928811
STEP: [reencrypt] Reading "/tmp/runConformanceTests912928811"
STEP: [reencrypt] Decoding "/tmp/runConformanceTests912928811"
STEP: Analyzing results
Mar 18 12:30:08.198: INFO: [reencrypt] TestCase ID: "http2/4.2.3" is a known failure; ignoring
[AfterEach] [sig-network-edge][Conformance][Area:Networking][Feature:Router]
  /home/aim/go-projects/origin/src/github.com/openshift/origin/_output/local/go/src/github.com/openshift/origin/test/extended/util/client.go:101
Mar 18 12:30:08.297: INFO: Deleted {user.openshift.io/v1, Resource=users  e2e-test-router-h2spec-8dl5l-user}, err: <nil>
Mar 18 12:30:08.394: INFO: Deleted {oauth.openshift.io/v1, Resource=oauthclients  e2e-client-e2e-test-router-h2spec-8dl5l}, err: <nil>
Mar 18 12:30:08.489: INFO: Deleted {oauth.openshift.io/v1, Resource=oauthaccesstokens  yG_tgwKRS1SnSLOXccW6VAAAAAAAAAAA}, err: <nil>
[AfterEach] [sig-network-edge][Conformance][Area:Networking][Feature:Router]
  /home/aim/go-projects/origin/src/github.com/openshift/origin/_output/local/go/src/github.com/openshift/origin/vendor/k8s.io/kubernetes/test/e2e/framework/framework.go:155
Mar 18 12:30:08.489: INFO: Waiting up to 7m0s for all (but 100) nodes to be ready
STEP: Destroying namespace "e2e-test-router-h2spec-8dl5l" for this suite.
[AfterEach] [sig-network-edge][Conformance][Area:Networking][Feature:Router]
  /home/aim/go-projects/origin/src/github.com/openshift/origin/_output/local/go/src/github.com/openshift/origin/test/extended/router/h2spec.go:52
Mar 18 12:30:08.856: INFO: Running AfterSuite actions on all nodes
Mar 18 12:30:08.856: INFO: Running AfterSuite actions on node 1

aim@spicy:~/go-projects/origin/src/github.com/openshift/origin
$ oc get namespaces | grep h2spec
e2e-test-router-h2spec-8dl5l                            Terminating   40s
```

Reply:
If you're not doing it in this test code, what's deleting the ns?!

Usually we must clean up after ourselves in these tests.